proposals/kubernetes-image-and-attestation-verification.md

```diff
@@ -0,0 +1,184 @@
+
+We fetch attestations (signed metadata) using [ORAS](https://pkg.go.dev/oras.land/oras-go/v2) Client stored with the image to run signature and data validation checks as defined by the policy. The notary signature verification also uses ORAS repository implementataion for its verification method. 
+
+For authentication with registry, we create a secret containing username password and server information and use that for authentication. We use [GCR Crane](https://pkg.go.dev/github.com/google/go-containerregistry/pkg/crane) for authentication using credentials.
+
+Our current Authentication system works but it is not perfect and as robust as other implementations. We support authentication using credentials but we dont have direct support for major OCI registry (Azure, AWS, GCP).
+
+Also we are dependent on ORAS as well as Crane at the same time. Both library serve the same purpose of communicating with OCI registries and have very similar methods.
+
+This Proposal proposes the following:
+1. Creating an inhouse implementation Notary repository which uses crane instead of ORAS to remove the indirect dependency on ORAS.
+2. Use crane instead of ORAS to fetch attestations and predecates.
+3. Update the authentication to support Azure, AWS, GCP or use [Flux Auth](https://pkg.go.dev/github.com/fluxcd/pkg/oci/auth@v0.23.0) package, which implements crane.
+
+# Implementation
+
+To update the Notary Repository to use crane, We will need to create a Repository client which implements the following interfact
+
+### Creating an inhouse implementation notary repository which uses crane
+```go
+type Repository interface {
+	// Resolve resolves a reference(tag or digest) to a manifest descriptor
+	Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error)
+
+	// ListSignatures returns signature manifests filtered by fn given the
+	// artifact manifest descriptor
+	ListSignatures(ctx context.Context, desc ocispec.Descriptor, fn func(signatureManifests []ocispec.Descriptor) error) error
+
+	// FetchSignatureBlob returns signature envelope blob and descriptor for
+	// given signature manifest descriptor
+	FetchSignatureBlob(ctx context.Context, desc ocispec.Descriptor) ([]byte, ocispec.Descriptor, error)
+
+	// PushSignature creates and uploads an signature manifest along with its
+	// linked signature envelope blob.
+	PushSignature(ctx context.Context, mediaType string, blob []byte, subject ocispec.Descriptor, annotations map[string]string) (blobDesc, manifestDesc ocispec.Descriptor, err error)
+}
+```
+All these methods in [Notary's implementation](https://github.com/notaryproject/notation-go/blob/v1.0.0-rc.4/registry/repository.go) currently use ORAS.
+Here is the implementation of all these functions using crane: https://github.com/Vishal-Chdhry/kyverno/blob/notary-v2-attestations/pkg/notary/repository.go
```

Comment:
It's best if you not link to your fork with a permalink including line numbers as these can change. Probably either put in a gist or in this proposal as a separate file.